dwave/optimization/symbols.pyx

```diff
@@ -2327,6 +2403,176 @@ cdef class NaryMultiply(ArraySymbol):
 _register(NaryMultiply, typeid(cppNaryMultiplyNode))
 
 
+# TODO: consider different location for this?
+class UnsupportedNaryReduceExpression(Exception):
+    def __init__(self, message: str, symbol: Symbol):
+        super().__init__(message)
+        self.symbol = symbol
+
+
+cdef class NaryReduce(ArraySymbol):
+    """Using a supplied :class:`~dwave.optimization.model.Expression`, perform
+    a reduction operation along one or more array operands. The reduction
+    operation (represented by the `Expression`) takes as input one value from
+    each of the operand arrays, as well as the result of the previously
+    computed operation, and computes a new value at the next output index.
+
+    This takes inspiration from
+    `numpy.ufunc.reduce <https://numpy.org/doc/2.1/reference/generated/numpy.ufunc.reduce.html>`_ 
+    but is different in that the reduction operation can take an arbitrary
+    number of arguments, instead of always two (hence the "n-ary" in
+    `NaryReduce`).
+
+    Examples:
+        This example performs a simple cumulative sum on an integer variable.
+
+        >>> from dwave.optimization.model import Expression, Model
+        >>> from dwave.optimization.symbols import NaryReduce
+        >>> model = Model()  # the main model
+        >>> x = model.integer(10, lower_bound=0, upper_bound=5)
+        >>> expr = Expression()  # the reduction operation
+        >>> xi = expr.input(0, 5, integral=True)  # will take the values of `x`
+        >>> # will take the previous result of the reduction, so the bounds must
+        >>> # include the entire possible range of the output of the reduction
+        >>> previous = expr.input(0, 5 * 10, integral=True)
+        >>> expr.set_output(xi + previous)
+        >>> cumulative_sum_x = NaryReduce(expr, (x,))
+        >>> type(cumulative_sum_x)
+        <class 'dwave.optimization.symbols.NaryReduce'>
+    """
+
+    def __init__(
+        self,
+        expression: Expression,
+        operands: Collection[ArraySymbol],
```

Comment:
These types don't do anything, so I'd rather remove them, and the `from typing import Collection`.